### DISCUSSION TITLE
Bad error messages

### DISCUSSION THREAD

**Discussion by weiznich** [tokio-rs/axum#1138] [General]
:wave:

As part of my rust foundation project grant I will work on ways to improve error messages generated by trait heavy crates.  I just want to reach out to collect examples of potential bad error messages emitted by rustc due to the (mis-) usage of axum. I would be happy to hear about such cases here or as [PR to this repo](https://github.com/weiznich/rust-foundation-community-grant)

**Comment by jplatte** [comment]
The worst ones definitely are when you try to [register a route handler](https://docs.rs/axum/0.5.10/axum/#routing), but it doesn't implement the `Handler` trait for some reason. We have a macro attribute that you can apply to the handler function to get a better error message: [`axum_macros::debug_handler`](https://docs.rs/axum-macros/0.2.3/axum_macros/attr.debug_handler.html). For a list of cases where a function doesn't implement this trait, and the correponding output from `debug_handler`, see [its trybuild fail tests](https://github.com/tokio-rs/axum/tree/8c31bee9bccf28720c3b5d89a1d4e554fcb0efff/axum-macros/tests/debug_handler/fail). I think the error message without `debug_handler` is always pretty much the same – just "the trait Handler<...> is not implemented for ..." without any more detail.
